### PR TITLE
New Storage Driver: ScyllaDb

### PR DESCRIPTION
**Pull request by Chris-Basson-Grundling**
Hey All.

This pr aims to add new Storage implementation of ScyllaDb.

Thanks for the reviews.

### CODE REVIEW

File: .github/workflows/test-scylladb.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+jobs:
+  Tests:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        go-version:
+          - 1.19.x
+          - 1.20.x
```
Comment (gaby) [inline .github/workflows/test-scylladb.yml:22]: Add 1.21.x here

File: scylladb/README.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
+---
+id: scylladb
+title: ScyllaDb
+---
+
+# ScyllaDb
+
+A ScyllaDb storage engine for [Fiber](github.com/gofiber/fiber) using [gocql](github.com/gocql/gocql).
```
Comment (gaby) [inline scylladb/README.md:8]: Is this the right lib? Gocql?

File: scylladb/README.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+func (s *Storage) Get(key string) ([]byte, error)
+func (s *Storage) Set(key string, value []byte, expire time.Duration) error
+func (s *Storage) Delete(key string) error
+func (s *Storage) Reset() error
+func (s *Storage) Close() error
+func (s *Storage) Conn() *gocql.Session
+```
+
+### Installation
+ScyllaDb is tested on the 2 last [Go versions](https://golang.org/dl/) with support for modules. So make sure to initialize one first if you didn't do that yet:
```
Comment (gaby) [inline scylladb/README.md:29]: If so remove 1.19.x from ci file

File: scylladb/go.mod [hunk 4]
```diff
@@ -0,0 +1,14 @@
+module github.com/gofiber/storage/scylladb
+
+go 1.19
```
Comment (gaby) [inline scylladb/go.mod:3]: Use 1.20.x here

File: scylladb/scylladb_test.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
+package scylladb
+
+import (
+	"github.com/gofiber/utils"
```
Comment (gaby) [inline scylladb/scylladb_test.go:4]: All the storage tests should use testify for Asserts. See here https://github.com/gofiber/storage/blob/main/redis/redis_test.go

**Review by gaby — CHANGES_REQUESTED** [review]

File: scylladb/README.md (hunk 2 above)
Comment (gaby) [reply scylladb/README.md:8]: According to their docs we should be using this https://github.com/scylladb/gocql

Source: https://opensource.docs.scylladb.com/stable/using-scylla/drivers/cql-drivers/scylla-go-driver.html

**Comment by ReneWerner87** [comment]
Pls add a section in the root README.md

File: scylladb/config.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
+package scylladb
+
+import "github.com/gocql/gocql"
+
+type Config struct {
```
Comment (gaby) [inline scylladb/config.go:5]: One thing that I see missing is support for setting a TLS config.

File: scylladb/config.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	//
+	// Optional. Default is nil
+	Session *gocql.Session
+
+	// Keyspace name
+	//
+	// Optional. Default is "fiber"
+	Keyspace string
+
+	// Host name where the ScyllaDb cluster is hosted
```
Comment (gaby) [inline scylladb/config.go:16]: Update the comment to indicate it supports multiple hosts

File: scylladb/go.mod [hunk 8]
```diff
@@ -0,0 +1,19 @@
+module github.com/gofiber/storage/scylladb
+
+go 1.20
+
+require (
+	github.com/gocql/gocql v1.6.0
+	github.com/gofiber/utils v1.1.0
```
Comment (gaby) [inline scylladb/go.mod:7]: Running `go mod tidy` should get rid of `gofiber/utils` 🤔 i dont see it being usef anywhere, right?

File: scylladb/config.go (hunk 6 above)
Comment (Chris-Basson-Grundling) [reply scylladb/config.go:7]: Added support for setting TLS config.

File: scylladb/go.mod [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+require (
+	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/golang/snappy v0.0.3 // indirect
+	github.com/hailocab/go-hostpool v0.0.0-20160125115350-e80d13ce29ed // indirect
+	github.com/pmezard/go-difflib v1.0.0 // indirect
+	gopkg.in/inf.v0 v0.9.1 // indirect
+)
+
+replace github.com/gocql/gocql => github.com/scylladb/gocql v1.11.1
```
Comment (gaby) [inline scylladb/go.mod:18]: I dont know if dependabot will track this, may be better to fix the require import at the top of the go.mod

File: scylladb/scylladb_test.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+	var (
+		key = "john"
+		val = []byte("doe")
+		exp = 1 * time.Second
+	)
+
+	err := testStore.Set(key, val, exp)
+	require.NoError(t, err)
+
+	time.Sleep(1100 * time.Millisecond)
```
Comment (gaby) [inline scylladb/scylladb_test.go:58]: There needs to be a `get()` after the sleep, to verify that the `key` expired after 1sec

**Review by gaby — CHANGES_REQUESTED** [review]
Overall looks good, one small change to the tests left

File: scylladb/go.mod [hunk 11]
```diff
@@ -2,6 +2,8 @@ module github.com/gofiber/storage/scylladb
 
 go 1.20
 
+replace github.com/gocql/gocql => github.com/scylladb/gocql v1.11.1
```
Comment (gaby) [inline scylladb/go.mod:5]: I mean the import on line 8 should be the one in the replace. I'm not sure if dependabot will pick it up as a "replace" directive

**Review by gaby — CHANGES_REQUESTED** [review]

File: scylladb/scylladb.go [hunk 12]
```diff
@@ -137,7 +137,8 @@ func (s *Storage) Get(key string) ([]byte, error) {
 		}
 		return nil, err
 	}
-	if expiration > 0 && time.Now().Unix() > expiration {
+	// If the expiration time has already passed, then return nil
+	if expiration != 0 && expiration <= time.Now().Unix() {
```
Comment (gaby) [inline scylladb/scylladb.go:141]: Does this mean that the entry stays in the DB? Maybe we should call delete instead once it's expired

**Review by gaby — CHANGES_REQUESTED** [review]

File: scylladb/scylladb.go (hunk 12 above)
Comment (Chris-Basson-Grundling) [reply scylladb/scylladb.go:141]: Great Point, not sure how I missed that, will update shortly.

File: scylladb/go.mod (hunk 11 above)
Comment (Chris-Basson-Grundling) [reply scylladb/go.mod]: From the tests it should be fine.
https://github.com/dependabot/dependabot-core/blob/main/go_modules/spec/fixtures/go_mods/go.mod

File: scylladb/scylladb.go (hunk 12 above)
Comment (Chris-Basson-Grundling) [reply scylladb/scylladb.go]: Updated to use default scylladb TTL

File: scylladb/scylladb.go [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+// New creates a new storage
+func New(config ...Config) *Storage {
+	var err error
+	var session *gocql.Session
+
+	// Set default config
+	cfg := configDefault(config...)
+
+	if cfg.Keyspace == "" {
+		panic(keyspaceMsg)
```
Comment (gaby) [inline scylladb/scylladb.go:40]: I don't think this ever happens, the configDefault() call above will set the Keyspace to the default value if empty. One thing that could happen is if the user set the Keyspace to something like "   ".

File: scylladb/config.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+	if cfg.Port <= 0 {
+		cfg.Port = ConfigDefault.Port
+	}
+	if cfg.Table == "" {
+		cfg.Table = ConfigDefault.Table
+	}
+	if cfg.Keyspace == "" {
+		cfg.Keyspace = ConfigDefault.Keyspace
+	}
+	if cfg.Consistency == "" {
```
Comment (gaby) [inline scylladb/config.go:97]: You can use:

```go
if len(strings.TrimSpace(yourVariable)) == 0 {
  // set default here
}
```

I just realized this needs to be fixed across all drivers. That's a different issue to deal with later. :-)

File: scylladb/scylladb.go (hunk 13 above)
Comment (gaby) [reply scylladb/scylladb.go:40]: Added a comment above to fix that issue in configDefault()

File: scylladb/scylladb.go [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	if cfg.Session == nil {
+		// Create a new cluster
+		cluster := gocql.NewCluster(cfg.Hosts...)
+		cluster.Consistency = gocql.ParseConsistency(cfg.Consistency)
+		cluster.Port = cfg.Port
+
+		// Set credentials if provided
+		if cfg.Username != "" && cfg.Password != "" {
```
Comment (gaby) [inline scylladb/scylladb.go:50]: This onen needs to be updated too

**Review by gaby — CHANGES_REQUESTED** [review]
Looks good overall and tests are passing. Just a small tweak leff to do.

File: .github/workflows/benchmark.yml [hunk 16]
```diff
@@ -117,6 +117,11 @@ jobs:
         run: |
           docker run -d --restart always -p 9000:9000 --name storage-minio -e MINIO_ROOT_USER='minio-user' -e MINIO_ROOT_PASSWORD='minio-password' minio/minio server /data
 
+      - name: Install ScyllaDb
+        run: |
+          docker run --name scylladb -p 9042:9042 -p 19042:19042 -p 9160:9160 -p 7000:7000 -p 7001:7001 -p 7199:7199 -p 9180:9180 -d scylladb/scylla:latest --broadcast-address 127.0.0.1 --listen-address 0.0.0.0 --broadcast-rpc-address 127.0.0.1
+          sleep 30 # Wait for ScyllaDb to initialize
```
Comment (efectn) [inline .github/workflows/benchmark.yml:123]: Can we reduce sleep time? Does 15 work stable?

File: scylladb/scylladb.go [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+	tableName string
+
+	selectQuery string
+	insertQuery string
+	deleteQuery string
+	resetQuery  string
+}
+
+var (
+	checkSchemaMsg = "the `value` row has an incorrect data type. " +
```
Comment (efectn) [inline scylladb/scylladb.go:22]: Can you create error vars with errors.New

File: scylladb/scylladb.go [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+	// Create keyspace if it does not exist
+	if err = session.Query(fmt.Sprintf(createKeyspaceQuery, cfg.Keyspace)).Exec(); err != nil {
+		session.Close()
+		panic(err)
+	}
+
+	// Drop table if reset is true
+	if cfg.Reset {
+		if err = session.Query(fmt.Sprintf(dropQuery, cfg.Keyspace, cfg.Table)).Exec(); err != nil {
```
Comment (efectn) [inline scylladb/scylladb.go:80]: Just curious. Do we get panic if bucket is empty?

File: scylladb/scylladb.go [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+			session.Close()
+			panic(err)
+		}
+	}
+
+	// Create the storage
+	store := &Storage{
+		session:     session,
+		tableName:   cfg.Table,
+		selectQuery: fmt.Sprintf("SELECT value FROM %s.%s WHERE key = ?", cfg.Keyspace, cfg.Table),
```
Comment (efectn) [inline scylladb/scylladb.go:90]: Can you make these unexported vars as well?

File: scylladb/scylladb.go (hunk 18 above)
Comment (Chris-Basson-Grundling) [reply scylladb/scylladb.go:85]: Yes, you will get the error message "gocql: no streams available on connection"
Maximum number of inflight requests allowed per connection defaults to 3276.

**Review by gaby — APPROVED** [review]
LGTM! Thanks for the PR

**Review by ReneWerner87 — APPROVED** [review]

**Pull request merged**